### PR TITLE
Adding Yul and Yul+ to smart contract languages

### PR DESCRIPTION
**Pull request by franzihei**
## Description

- Adding Yul and Yul+ to smart contract languages
- Adding material and descriptions for Yul and Yul+
- Adjusting intro text accordingly

Example smart contract in Yul is missing, I'll reach out to the team and add it later.

### CODE REVIEW

**Comment by ryancreatescopy** [comment]
@franzihei – loving these PRs, you're on :fire:

Will merge once we have the example contract in there!

File: src/content/developers/docs/smart-contracts/languages/index.md
Comment (chriseth) [inline src/content/developers/docs/smart-contracts/languages/index.md:207]:
```suggestion
- Supports EVM and eWASM, and it is designed to be a usable common denominator of both platforms.
```

File: src/content/developers/docs/smart-contracts/languages/index.md
Comment (chriseth) [inline src/content/developers/docs/smart-contracts/languages/index.md:240]: I would add:
```
let result := power(calldataload(0), calldataload(32))
mstore(0, result)
return(0, 32)
```
otherwise the optimizer will completely remove the code.

File: src/content/developers/docs/smart-contracts/languages/index.md
Comment (chriseth) [inline src/content/developers/docs/smart-contracts/languages/index.md:244]: maybe also link to `latest`?

File: src/content/developers/docs/smart-contracts/languages/index.md
Comment (chriseth) [inline src/content/developers/docs/smart-contracts/languages/index.md:263]:
```suggestion
- Simplistic and functional low-level language. 
```
why grammar?

File: src/content/developers/docs/smart-contracts/languages/index.md
Comment (chriseth) [inline src/content/developers/docs/smart-contracts/languages/index.md:223]: Maybe add `--bin` because otherwise it will not output anything. Or provide a full commandline:
``solc --strict-assembly --bin input.yul``
and meniton taht the example should be stored in the file `input.yul`

File: src/content/developers/docs/smart-contracts/languages/index.md
Comment (chriseth) [inline src/content/developers/docs/smart-contracts/languages/index.md:226]:
```suggestion
```

File: src/content/developers/docs/smart-contracts/languages/index.md
Comment (franzihei) [reply src/content/developers/docs/smart-contracts/languages/index.md:239]: add it in which line?

File: src/content/developers/docs/smart-contracts/languages/index.md
Comment (ryancreatescopy) [inline src/content/developers/docs/smart-contracts/languages/index.md:204]: I think it might be nice to have a short introduction to these two languages. I realise that there's an intro right at the top of the page but I feel like a lot of people would miss this.

File: src/content/developers/docs/smart-contracts/languages/index.md
Comment (ryancreatescopy) [inline src/content/developers/docs/smart-contracts/languages/index.md:207]: Can we add an explainer of what eWASM refers to and link to EVM page?

File: src/content/developers/docs/smart-contracts/languages/index.md
Comment (ryancreatescopy) [inline src/content/developers/docs/smart-contracts/languages/index.md:208]: I'm not following this but I am also not a developer. Will defer to @samajammin here, but we may want to explain what this means a little more.

File: src/content/developers/docs/smart-contracts/languages/index.md
Comment (ryancreatescopy) [inline src/content/developers/docs/smart-contracts/languages/index.md:212]: Can we link to layer 2 scaling page here: `/en/developers/docs/layer-2-scaling/`

File: src/content/developers/docs/smart-contracts/languages/index.md
Comment (ryancreatescopy) [inline src/content/developers/docs/smart-contracts/languages/index.md:221]: Excuse my ignorance but what does this contract do? We should maybe explain what this simple example does.

**Review by ryancreatescopy — CHANGES_REQUESTED** [review]

File: src/content/developers/docs/smart-contracts/languages/index.md
Comment (franzihei) [reply src/content/developers/docs/smart-contracts/languages/index.md:204]: Yeah but then we basically need to repeat the same all over again --> Yul is an intermediate language for Ethereum. There's not much more to say than the stuff that is mentioned in the bullet points below.

Comment (ryancreatescopy) [reply src/content/developers/docs/smart-contracts/languages/index.md:204]: Why would someone want to use Yul or Yul+ ? After a Google, I came across this description of Yul+ "It allows the developer to get much closer to raw EVM than Solidity, and with that comes the promise to drastically improve gas usage."

Could we say something like "Yul gives you more control for optimising your smart contract, like reducing the amount of Gas it needs and..."

What are the other outcomes a developer might be able to achieve by using Yul instead of Solidity – hope this makes sense.

File: src/content/developers/docs/smart-contracts/languages/index.md
Comment (franzihei) [reply src/content/developers/docs/smart-contracts/languages/index.md:221]: It's just a super easy power on/off contract that doesn't "do" anything useful, just shows the language. For the more useful stuff the ERC20 contract is interesting, but very long. I don't think we need to explain it in more detail here since the people that Yul is interesting to _should_ be able to get what they need from it. Maybe we should just move the "disclaimer" that I put in the comparison section right on top of the Yul/Yul+ section so that it becomes clear that this is not necessarily for beginners? --> **If you're new to Ethereum and you haven't done any coding with smart contract languages yet, we recommend getting started with Solidity or Vyper and, only once you're familiar with smart contract security best practices and the specifics of working with the EVM, looking into Yul or Yul+.**

File: src/content/developers/docs/smart-contracts/languages/index.md
Comment (franzihei) [reply src/content/developers/docs/smart-contracts/languages/index.md:204]: It does make sense but then we're repeating what is said in the comparison section. :) There is also no such "arguments" in the Vyper or Solidity sections (just to be consistent).

Comment (franzihei) [reply src/content/developers/docs/smart-contracts/languages/index.md:204]: Also, the "Intro to Yul+" post that you googled is already linked below in the important resources section. :)

Comment (franzihei) [reply src/content/developers/docs/smart-contracts/languages/index.md:204]: I was trying to follow the structure of the page and put advantages of the language in the comparison section and the objective traits of the lang in the introductory bullet points, just like it's being done with the other two languages.

File: src/content/developers/docs/smart-contracts/languages/index.md
Comment (chriseth) [inline src/content/developers/docs/smart-contracts/languages/index.md:239]: Maybe we should use if now (didn't exist when the example was written):
```suggestion
            if mod(exponent, 2) { result := mul(base, result) }
```

File: src/content/developers/docs/smart-contracts/languages/index.md
Comment (chriseth) [reply src/content/developers/docs/smart-contracts/languages/index.md:242]: Above the last `}`

File: src/content/developers/docs/smart-contracts/languages/index.md
Comment (samajammin) [reply src/content/developers/docs/smart-contracts/languages/index.md:210]: I'm also unclear on this @franzihei - mostly around what "target" is referring to.
- Does "Good target for high-level optimisation stages" == "Useful language for high-level optimisation stages"?
- Does "all target platforms"== "EVM and eWASM"?

File: src/content/developers/docs/smart-contracts/languages/index.md
Comment (samajammin) [inline src/content/developers/docs/smart-contracts/languages/index.md:210]:
```suggestion
- Good target for high-level optimisation stages that can benefit both EVM and eWASM platforms equally.
```

File: src/content/developers/docs/smart-contracts/languages/index.md
Comment (samajammin) [inline src/content/developers/docs/smart-contracts/languages/index.md:225]:
```suggestion
The following simple example implements a power function. It can be compiled using ``solc --strict-assembly --bin input.yul``. The example should
```

**Review by samajammin — DISMISSED** [review]
@franzihei @ryancreatescopy making a couple edits based on the PR discussions. Please let me know if I misinterpreted anything. I welcome any further edits but am going to merge this in for now! Thanks!

**Review by samajammin — APPROVED** [review]

**Comment by samajammin** [comment]
@all-contributors please add @chriseth for review.

**Pull request merged**
